### ISSUE TITLE
Can/should `is` operator update null-state in false case?

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#30297]
```C#
string? s = null;
if (!(s is object)) { s.ToString(); /* could warn? */ }
if (!(s is string)) { s.ToString(); /* could warn? */ }
```

There are variants of this scenario with `string!` and `string~`. The question is whether we should learn in both branches, or whether we should treat one branch as unreachable.

Relates to https://github.com/dotnet/roslyn/pull/30283

**Comment by cston** [comment]
We should have similar behavior for `==`:
```C#
    static void Main()
    {
        string? s = "";
        if (s == null)
            _ = s.Length; // should warn?
    }
```

**Comment by jcouv** [comment]
This was clarified in LDM and is tracked by https://github.com/dotnet/roslyn/issues/33526

In short, `s == null` is a "pure" null test, so affects both branches. But `s is object` is not.
